src/libs/sfdk/dockervirtualmachine.cpp

```diff
@@ -0,0 +1,531 @@
+        arguments.append(PUBLISH_PORT);
+        arguments.append(QString::number(hostPort) + ":" + QString::number(guestPort));
+    };
+
+    forwardPort(cachedInfo().sshPort, GUESTSSHPORT);
+    forwardPort(cachedInfo().wwwPort, GUESTWWWPORT);
+
+    auto sharedPath = [&arguments](const QString& hostPath, const QString& guestPath) {
+        arguments.append(VOLUME);
+        arguments.append(hostPath + ":" + guestPath);
+    };
+
+    sharedPath(cachedInfo().sharedHome, Constants::BUILD_ENGINE_GUEST_SHARED_HOME);
+    sharedPath(cachedInfo().sharedSrc, Constants::BUILD_ENGINE_GUEST_SHARED_SRC);
+    sharedPath(cachedInfo().sharedTargets, Constants::BUILD_ENGINE_GUEST_SHARED_TARGET);
+    sharedPath(cachedInfo().sharedConfig, Constants::BUILD_ENGINE_GUEST_SHARED_CONFIG);
+    sharedPath(cachedInfo().sharedSsh, Constants::BUILD_ENGINE_GUEST_SHARED_SSH);
+
+    arguments.append(q->name());
+
+    auto runner = std::make_unique<ProcessRunner>(dockerPath(), arguments);
+    QObject::connect(runner.get(), &ProcessRunner::done, q,
+                     [q, functor, process = runner->process()](bool ok) {
+        if (ok)
+            q->m_containerId = QString::fromLocal8Bit(process->readAllStandardOutput().trimmed());
+        functor(ok);
+    });
+    SdkPrivate::commandQueue()->enqueue(std::move(runner));
+}
+
+void DockerVirtualMachinePrivate::stop(const QObject *context, const Functor<bool> &functor)
+{
+    Q_Q(DockerVirtualMachine);
+    Q_ASSERT(context);
+    Q_ASSERT(functor);
+
+    const QPointer<const QObject> context_{context};
+
+    auto enqueue = [=](const QStringList &arguments, CommandQueue::BatchId batch) {
+        auto runner = std::make_unique<ProcessRunner>(dockerPath(), arguments);
+        QObject::connect(runner.get(), &ProcessRunner::failure, Sdk::instance(), [=]() {
+            SdkPrivate::commandQueue()->cancelBatch(batch);
+            callIf(context_, functor, false);
+        });
+        ProcessRunner *runner_ = runner.get();
+        SdkPrivate::commandQueue()->enqueue(std::move(runner));
+        return runner_;
+    };
+
+    const CommandQueue::BatchId batch = SdkPrivate::commandQueue()->beginBatch();
+
+    QStringList stopArguments;
+    stopArguments.append(STOP);
+    stopArguments.append(q->m_containerId);
+
+    enqueue(stopArguments, batch);
+
+    QStringList commitArguments;
+    commitArguments.append(COMMIT);
+    commitArguments.append(q->m_containerId);
+    commitArguments.append(q->name());
+
+    enqueue(commitArguments, batch);
+
+    SdkPrivate::commandQueue()->enqueueCheckPoint(context, [=]() { functor(true); });
+    SdkPrivate::commandQueue()->endBatch();
+}
+
+void DockerVirtualMachinePrivate::probe(const QObject *context,
+        const Functor<VirtualMachinePrivate::BasicState, bool> &functor) const
+{
+    Q_Q(const DockerVirtualMachine);
+    Q_ASSERT(context);
+    Q_ASSERT(functor);
+
+    fetchRegisteredVirtualMachines(q, [=](const QStringList &registeredVms, bool ok) {
```

Comment:
Using now <code>context</code> as context, as there is no longer any reason to use <code>q</code>.

Reply:
Unfortunately it is not that easy. ``m_container` is still accessed there and that needs to be guarded with `q`. Even if it can be accessed directly now, its lifetime is bound to `q`. Wouldn't it be possible to initialize `m_containerId` from `fetchInfo` instead to get rid of that?

Reply:
We don't know the container id in fetchInfo, as that method only deals with images (not containers)